core/src/main/java/org/apache/druid/segment/column/Types.java

```diff
@@ -112,4 +121,529 @@
+
+      buffer.put(offset, NullHandling.IS_NOT_NULL_BYTE);
+      buffer.putInt(offset + 1, array.length);
+      for (Long element : array) {
+        if (element != null) {
+          checkMaxBytes(
+              ColumnType.LONG_ARRAY,
+              sizeBytes + 1 + Long.BYTES,
+              maxSizeBytes
+          );
+          sizeBytes += writeNullableLong(buffer, offset + sizeBytes, element);
+        } else {
+          checkMaxBytes(
+              ColumnType.LONG_ARRAY,
+              sizeBytes + 1,
+              maxSizeBytes
+          );
+          sizeBytes += writeNull(buffer, offset + sizeBytes);
+        }
+      }
+      return sizeBytes;
+    }
+  }
+
+  /**
+   * Reads a nullable variably lengthed Long[] value from a {@link ByteBuffer} at the supplied offset. If the null byte
+   * is set to {@link NullHandling#IS_NULL_BYTE}, this method will return null, else it will read the size of the array
+   * from the next 4 bytes and then read that many elements with {@link #isNullableNull} and {@link #readNullableLong}.
+   *
+   * layout: | null (byte) | size (int) | {| null (byte) | long |, | null (byte) |, ... |null (byte) | long |} |
+   *
+   * This method does not change the buffer position, limit, or mark.
+   */
+  @Nullable
+  public static Long[] readNullableLongArray(ByteBuffer buffer, int offset)
+  {
+    // | null (byte) | array length (int) | array bytes |
+    if (isNullableNull(buffer, offset++)) {
+      return null;
+    }
+    final int longArrayLength = buffer.getInt(offset);
+    offset += Integer.BYTES;
+    final Long[] longs = new Long[longArrayLength];
+    for (int i = 0; i < longArrayLength; i++) {
+      if (isNullableNull(buffer, offset)) {
+        longs[i] = null;
+      } else {
+        longs[i] = readNullableLong(buffer, offset);
+        offset += Long.BYTES;
+      }
+      offset++;
+    }
+    return longs;
+  }
+
+  /**
+   * Write a variably lengthed Double[] value to a {@link ByteBuffer} at the supplied offset. The first byte is set to
+   * {@link NullHandling#IS_NULL_BYTE} or {@link NullHandling#IS_NOT_NULL_BYTE} as appropriate, and if the Long[] value
+   * is not null, the size in bytes is written as an integer in the next 4 bytes. Elements of the array are each written
+   * out with {@link #writeNull} if null, or {@link #writeNullableDouble} if not, taking either 1 or 9 bytes each. If
+   * the total byte size of serializing the array is larger than the max size parameter, this method will explode via a
+   * call to {@link #checkMaxBytes}.
+   *
+   * layout: | null (byte) | size (int) | {| null (byte) | double |, | null (byte) |, ... |null (byte) | double |} |
+   *
+   *
+   * This method does not permanently change the buffer position, limit, or mark.
+   *
+   * @return number of bytes written (1 if null, or 5 + size of Double[] if not)
+   */
+  public static int writeNullableDoubleArray(ByteBuffer buffer, int offset, @Nullable Double[] array, int maxSizeBytes)
```

Comment:
it is easier to be consistent across all types working in bytes, because variably sized types like strings and string arrays become sort of hard to reason about. The main reason here is that the expression aggregator has a single max size bytes parameter (since that aggregator can potentially aggregate any type)